### ISSUE TITLE
False negative in Style/BlockDelimiters when using semantic style in factories

### ISSUE THREAD

**Issue by ignatiusreza** [rubocop-hq/rubocop#4239]
Factory Girl allowed us to define [dynamic attributes](https://github.com/thoughtbot/factory_girl/blob/master/GETTING_STARTED.md#dynamic-attributes) by passing in lazy-evaluated block into attribute.. for example (taken directly from [their documentation](https://github.com/thoughtbot/factory_girl/blob/master/GETTING_STARTED.md))

```rb
factory :user do
  # ...
  activation_code { User.generate_activation_code }
  date_of_birth   { 21.years.ago }
end
```

the return value of these blocks are significant, as it is will be used as the value of the attributes, so they should be considered functional block.. the attributes provided by `Style/BlockDelimiters` -- `FunctionalMethods`, etc -- unfortunately won't help us here, since these attributes are dynamic..

I'm not so sure if it's something that we can safely analyze statically.. though, just to throw out ideas, is it possible to expand `FunctionalMethods` to allow defining nested blocks.. for example, to have something like `factory *`, or regex style like `factory (?!(before|after))`..

--------

## Expected behavior

Dynamic Attributes under `factory` should be recognized as functional block

## Actual behavior

Dynamic Attributes under `factory` is recognized as procedural block

## Steps to reproduce the problem

Define a factory like:

```rb
factory :user do
  # ...
  activation_code { User.generate_activation_code }
  date_of_birth   { 21.years.ago }
end
```

## RuboCop version

o.46.0

**Comment by bbatsov** [comment]
As it's really hard to reliably handle cases like this one, I really don't know what to suggest. We might implement some checks for nested blocks, but how would result in some false negatives in the end - e.g. there might be some procedural block in these contexts that won't be properly flagged. Once again - those are the limitations of dynamic languages. :-)

**Comment by ignatiusreza** [comment]
how do you feel about extending `FunctionalMethods` and `ProceduralMethods` config to support nested declaration like:

```yaml
ProceduralMethods:
  - tap
  - with_object
  - factory:
      MethodMissing: functional
      FunctionalMethods:
        - add_attribute
        - initialize_with
        - sequence
      ProceduralMethods:
        - callback
        - before
        - before_stubbed
        - after
        - after_stubbed
        - transient
        - to_create
```
if this is something that seems acceptable API-wise, I would love to try to take a stab at making a PR for it.. :grin:

**Comment by bbatsov** [comment]
There hasn't been much activity on this ticket and our Core Team is spread too thin on so many tasks, so I'll just close it for the sake of having a cleaner lists of tasks to focus on. We'd gladly review a PR, but it's unlikely that'd we're going to tackle this ourselves in the foreseeable future.

**Comment by ignatiusreza** [comment]
sure! thanks a lot, i'll try to come up with a PR so that we can have more concrete discussion..
